index.js

```diff
@@ -298,68 +370,105 @@ function cmpType (a, b) {
   return typeOrder[a.type] - typeOrder[b.type]
 }
 
-DB.prototype._onpt = function (pt, seen, cb) {
+// Given a node by its version, this collects the node itself, and also
+// recursively climbs all ways and relations that the node (or its referers)
+// are referred to by.
+// OsmVersion, { OsmVersion: Boolean } -> [OsmDocument]
+DB.prototype._collectNodeAndReferers = function (version, seenAccum, cb) {
   cb = once(cb || noop)
   var self = this
-  var link = pt.value.toString('hex')
-  if (has(seen, link)) return cb(null, [])
-  seen[link] = true
-  var res = [], added = {}, pending = 2
-  self.log.get(link, function (err, doc) {
-    if (doc && doc.value && doc.value.k && doc.value.v) {
-      addDoc(doc.value.k, link, doc.value.v)
-    }
-    if (--pending === 0) cb(null, res)
+  if (has(seenAccum, version)) return cb(null, [])
+  var res = [], added = {}, pending = 1
+
+  // Track the original node that came from the kdb query that brought us here,
+  // but don't add it yet. There are certain conditions (e.g. there is only one
+  // way containing us, and it's deleted) where the node would not be returned.
+  var selfNode
+  var originalNode
+  self.log.get(version, function (err, node) {
```

Comment:
Done. The reason why there's no error handling in some of these places is that the logic is very complex and it's not immediately obvious what the "right thing" to do is when an error crops up. We probably want to skip errors where recovery is possible (with good logging), and terminate everything on critical errors. Determining which is which and what to do seems like a big and complex enough task for a wholly separate PR though.